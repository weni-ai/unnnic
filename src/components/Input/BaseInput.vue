<template>
  <input
    v-if="mask"
    v-mask="mask"
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :readonly="readonly"
  />
  <input
    v-else
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :maxlength="maxlength"
    :readonly="readonly"
  />
</template>

<script>
import { mask } from 'vue-the-mask';
import { fullySanitize } from '../../utils/sanitize';

export default {
  directives: { mask },
  props: {
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: 'text',
    },
    size: {
      type: String,
      default: 'md',
    },
    mask: {
      type: [String, Array],
      default: '',
    },
    hasIconLeft: Boolean,
    hasIconRight: Boolean,
    hasClearIcon: Boolean,
    maxlength: {
      type: Number,
      default: null,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    forceActiveStatus: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['update:modelValue'],
  data() {
    return {};
  },
  computed: {
    attributes() {
      return {
        ...this.$attrs,
        ...this.$attrs['v-bind'],
        onInput: (event) => {
          this.$emit('update:modelValue', event.srcElement.value);
        },
      };
    },

    classes() {
      return [
        'input',
        `size-${this.size}`,
        this.type,
        {
          'input--has-icon-left': this.hasIconLeft,
          'input--has-icon-right': this.hasIconRight,
          'input--has-clear-icon': this.hasClearIcon,
        },
      ];
    },
  },
  methods: {
    fullySanitize,
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;
@use './Input.scss' as *;

.input {
  @include input-base;

  box-sizing: border-box;
  width: 100%;

  &.size {
    &-md {
      @include input-md-font;
      padding: $unnnic-space-3 $unnnic-space-4;
    }

    &-sm {
      @include input-sm-font;
      padding: $unnnic-space-2 $unnnic-space-4;
    }
  }

  &.input--has-icon-left {
    padding-left: $unnnic-space-10;
  }

  &.input--has-icon-right {
    padding-right: $unnnic-space-10;
    &.input--has-clear-icon {
      padding-right: $unnnic-space-10 + $unnnic-space-6;
    }
  }

  &.input--has-clear-icon {
      padding-right: $unnnic-space-10;
    }

  &.error {
    @include input-error;
  }

  &:-ms-input-placeholder,
  &::-ms-input-placeholder {
    /* Internet Explorer 10-11 */
    color: $unnnic-color-fg-muted;
  }
}
</style>
