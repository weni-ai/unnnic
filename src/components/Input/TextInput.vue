<template>
  <div
    :class="[
      'text-input',
      `size--${size}`,
      `text-input--icon-right-size-${iconRightSize}`,
    ]"
  >
    <BaseInput
      v-bind="attributes"
      ref="base-input"
      :modelValue="modelValue"
      :size="size"
      :nativeType="
        nativeType === 'password' && showPassword ? 'text' : nativeType
      "
      :type="type"
      class="input-itself"
      :hasIconLeft="!!iconLeft"
      :hasIconRight="!!iconRight || allowTogglePassword"
      @focus="onFocus"
      @blur="onBlur"
    />

    <UnnnicIcon
      v-if="iconLeft"
      :scheme="iconScheme"
      :icon="iconLeft"
      size="sm"
      :clickable="iconLeftClickable"
      :class="['icon-left', { clickable: iconLeftClickable }]"
      @click="onIconLeftClick"
    />

    <UnnnicIcon
      v-if="iconRightSvg"
      :scheme="iconScheme"
      :icon="iconRightSvg"
      :size="iconRightSize"
      :clickable="iconRightClickable || allowTogglePassword"
      :class="[
        'icon-right',
        { clickable: iconRightClickable || allowTogglePassword },
      ]"
      @click="onIconRightClick"
    />
  </div>
</template>

<script>
import BaseInput from './BaseInput.vue';
import UnnnicIcon from '../Icon.vue';

export default {
  components: {
    BaseInput,
    UnnnicIcon,
  },
  props: {
    placeholder: {
      type: String,
      default: null,
    },
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: '',
    },
    iconLeft: {
      type: String,
      default: null,
    },
    iconRight: {
      type: String,
      default: null,
    },
    iconLeftClickable: {
      type: Boolean,
      default: null,
    },
    iconRightClickable: {
      type: Boolean,
      default: null,
    },
    iconRightSize: {
      type: String,
      default: 'sm',
    },
    allowTogglePassword: {
      type: Boolean,
      default: null,
    },
    hasCloudyColor: {
      type: Boolean,
      default: null,
    },
    size: {
      type: String,
      default: 'md',
    },
  },
  emits: ['icon-left-click', 'icon-right-click'],
  data() {
    return {
      isFocused: false,
      showPassword: false,
    };
  },
  computed: {
    isDisabled() {
      return this.$attrs.disabled;
    },

    iconRightSvg() {
      if (this.allowTogglePassword) {
        return this.showPassword ? 'view-off-1' : 'view-1-1';
      }

      return this.iconRight;
    },

    iconScheme() {
      if (this.type === 'error') {
        return 'aux-red-500';
      }

      if (this.isDisabled) {
        return 'neutral-cleanest';
      }

      if (this.modelValue || this.isFocused) {
        return 'neutral-dark';
      }

      if (this.hasCloudyColor) {
        return 'neutral-cloudy';
      }

      return 'neutral-cloudy';
    },

    attributes() {
      return { ...this.$attrs, ...this.$attrs['v-bind'], ...this.$props };
    },
  },

  methods: {
    focus() {
      this.$refs['base-input'].$el.focus();
    },

    onFocus() {
      this.isFocused = true;
    },

    onBlur() {
      this.isFocused = false;
    },

    onIconLeftClick() {
      if (this.iconLeftClickable) this.$emit('icon-left-click');
    },

    onIconRightClick() {
      if (this.attributes.disabled) return;
      if (this.allowTogglePassword) this.showPassword = !this.showPassword;
      else if (this.iconRightClickable) this.$emit('icon-right-click');
    },
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;

.icon {
  &-left,
  &-right {
    &:not(.clickable) {
      pointer-events: none;
    }
  }

  &-left {
    position: absolute;
    top: $unnnic-spacing-ant + 0.1875 * $unnnic-font-size;
    left: $unnnic-inline-sm - $unnnic-border-width-thinner;
  }

  &-right {
    position: absolute;
    top: $unnnic-spacing-ant + 0.1875 * $unnnic-font-size;
    right: $unnnic-inline-sm - $unnnic-border-width-thinner;
  }
}
.text-input {
  position: relative;

  &.size--sm {
    .icon-left,
    .icon-right {
      top: $unnnic-spacing-xs + 0.125 * $unnnic-font-size;
    }
  }

  &.size--sm.text-input--icon-right-size-ant .icon-right {
    top: 0.5625 * $unnnic-font-size;
  }

  &.size--md.text-input--icon-right-size-ant .icon-right {
    top: 0.8125 * $unnnic-font-size;
  }
}
</style>
