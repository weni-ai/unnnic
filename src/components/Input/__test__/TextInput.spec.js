import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test, vi } from 'vitest';
import TextInput from '@/components/Input/TextInput.vue';
import BaseInput from '@/components/Input/BaseInput.vue';
import UnnnicIcon from '@/components/Icon.vue';

describe('TextInput.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(TextInput, {
      props: {
        placeholder: 'Enter text',
        type: 'normal',
        modelValue: '',
        nativeType: 'text',
        iconLeft: 'search',
        iconRight: 'clear',
        iconLeftClickable: true,
        iconRightClickable: true,
        allowTogglePassword: false,
        hasCloudyColor: false,
        size: 'md',
      },
      global: {
        components: {
          BaseInput,
          UnnnicIcon,
        },
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders BaseInput component', () => {
    const baseInput = wrapper.findComponent(BaseInput);
    expect(baseInput.exists()).toBe(true);
  });

  test('renders UnnnicIcon components when icons are provided', () => {
    const iconLeft = wrapper.find('.icon-left');
    const iconRight = wrapper.find('.icon-right');
    expect(iconLeft.exists()).toBe(true);
    expect(iconRight.exists()).toBe(true);
  });

  test('does not render UnnnicIcon components when icons are not provided', async () => {
    await wrapper.setProps({ iconLeft: null, iconRight: null });
    const iconLeft = wrapper.find('.icon-left');
    const iconRight = wrapper.find('.icon-right');
    expect(iconLeft.exists()).toBe(false);
    expect(iconRight.exists()).toBe(false);
  });

  test('applies the correct size class', async () => {
    await wrapper.setProps({ size: 'sm' });
    expect(wrapper.classes()).toContain('size--sm');
  });

  test('computes the correct iconRightSvg when allowTogglePassword is true', async () => {
    await wrapper.setProps({ allowTogglePassword: true });
    expect(wrapper.vm.iconRightSvg).toBe('view-1-1');

    await wrapper.setData({ showPassword: true });
    expect(wrapper.vm.iconRightSvg).toBe('view-off-1');
  });

  test('computes the correct iconScheme based on various states', async () => {
    await wrapper.setProps({ type: 'error' });
    expect(wrapper.vm.iconScheme).toBe('aux-red-500');

    await wrapper.setProps({ type: 'normal' });
    await wrapper.setData({ isDisabled: true });
    expect(wrapper.vm.iconScheme).toBe('neutral-cloudy');

    await wrapper.setData({ isDisabled: false });
    await wrapper.setProps({ modelValue: 'text' });
    expect(wrapper.vm.iconScheme).toBe('neutral-dark');

    await wrapper.setData({ isFocused: true });
    expect(wrapper.vm.iconScheme).toBe('neutral-dark');

    await wrapper.setProps({ hasCloudyColor: true, modelValue: '' });
    await wrapper.setData({ isFocused: false });
    expect(wrapper.vm.iconScheme).toBe('neutral-cloudy');
  });

  test('attributes computed property returns combined attributes and props', () => {
    const attributes = wrapper.vm.attributes;
    expect(attributes.placeholder).toBe('Enter text');
    expect(attributes.type).toBe('normal');
    expect(attributes.nativeType).toBe('text');
  });

  test('toggles password visibility when allowTogglePassword is true', async () => {
    await wrapper.setProps({
      nativeType: 'password',
      allowTogglePassword: true,
    });
    expect(wrapper.vm.nativeType).toBe('password');

    await wrapper.find('.icon-right').trigger('click');
    expect(wrapper.vm.showPassword).toBe(true);
    expect(
      wrapper.findComponent({ name: 'BaseInput' }).attributes('type'),
    ).toBe('text');

    await wrapper.find('.icon-right').trigger('click');
    expect(wrapper.vm.showPassword).toBe(false);
    expect(wrapper.vm.nativeType).toBe('password');
  });

  test('emits icon-left-click when left icon is clicked and clickable', async () => {
    const iconLeft = wrapper.find('.icon-left');
    await iconLeft.trigger('click');
    expect(wrapper.emitted('icon-left-click')).toBeTruthy();
  });

  test('emits icon-right-click when right icon is clicked and clickable', async () => {
    const iconRight = wrapper.find('.icon-right');
    await iconRight.trigger('click');
    expect(wrapper.emitted('icon-right-click')).toBeTruthy();
  });

  test('focus method calls focus on base input element', () => {
    const focusSpy = vi.spyOn(wrapper.vm.$refs['base-input'].$el, 'focus');
    wrapper.vm.focus();

    expect(focusSpy).toHaveBeenCalled();
    focusSpy.mockRestore();
  });

  test('onFocus method sets isFocused to true', async () => {
    expect(wrapper.vm.isFocused).toBe(false);

    wrapper.vm.onFocus();
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isFocused).toBe(true);
  });

  test('onBlur method sets isFocused to false', async () => {
    wrapper.setData({ isFocused: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isFocused).toBe(true);

    wrapper.vm.onBlur();
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isFocused).toBe(false);
  });

  test('UnnnicIcon clickable prop is correctly set based on iconRightClickable and allowTogglePassword', async () => {
    await wrapper.setProps({
      iconRightClickable: true,
      allowTogglePassword: false,
    });
    expect(wrapper.findAllComponents(UnnnicIcon)[1].props('clickable')).toBe(
      true,
    );

    await wrapper.setProps({
      iconRightClickable: false,
      allowTogglePassword: true,
    });
    expect(wrapper.findAllComponents(UnnnicIcon)[1].props('clickable')).toBe(
      true,
    );

    await wrapper.setProps({
      iconRightClickable: false,
      allowTogglePassword: false,
    });

    expect(wrapper.findAllComponents(UnnnicIcon)[1].props('clickable')).toBe(
      false,
    );

    await wrapper.setProps({
      iconRightClickable: true,
      allowTogglePassword: true,
    });
    expect(wrapper.findAllComponents(UnnnicIcon)[1].props('clickable')).toBe(
      true,
    );
  });

  test('onIconRightClick method does nothing if the input is disabled', async () => {
    await wrapper.setData({ attributes: { disabled: true } });

    const iconRightClickSpy = vi.spyOn(wrapper.vm, 'onIconRightClick');

    wrapper.vm.onIconRightClick();

    expect(iconRightClickSpy).toHaveBeenCalled();
    expect(wrapper.vm.showPassword).toBe(false);

    iconRightClickSpy.mockRestore();
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
