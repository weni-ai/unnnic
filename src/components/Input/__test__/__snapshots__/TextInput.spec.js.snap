// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput.vue > matches the snapshot 1`] = `
"<div data-v-a0d36167="" class="text-input size--md"><input data-v-86533b41="" data-v-a0d36167="" placeholder="Enter text" iconleft="search" iconright="clear" iconleftclickable="true" iconrightclickable="true" allowtogglepassword="false" hascloudycolor="false" showclear="false" class="input-itself input size-md normal input--has-icon-left input--has-icon-right input-itself" type="text" value=""><span data-v-26446d8e="" data-v-a0d36167="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--fg-base unnnic-icon-size--ant unnnic-icon__size--ant unnnic--clickable icon-left clickable" data-testid="material-icon" translate="no">search</span>
  <section data-v-a0d36167="" class="icon-right-container">
    <!--v-if--><span data-v-26446d8e="" data-v-a0d36167="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--fg-base unnnic-icon-size--ant unnnic-icon__size--ant unnnic--clickable icon-right clickable" data-testid="material-icon" translate="no">clear</span>
  </section>
</div>"
`;
